designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,308 @@
+- Standardize how both developers and end-users can reason about default rule options
+- Encourage writing rules that allow tooling such as [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator) to describe rule options programmatically - and therefore more consistently
+
+Options provided to rules would be the result of a deep (recursive) clone and merge of the rule's default options and user-provided options.
+
+## Detailed Design
+
+Currently, ESLint rule options are passed through a [`getRuleOptions` function](https://github.com/eslint/eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/lib/linter/linter.js#L740) whose only processing is to remove severity.
+That function could be augmented to also take in the `rule.meta.defaultOptions`, if it exists.
+
+Defaulting logic could go with the following logic to approximate user intent:
+
+1. If the user-provided value is `undefined`, go with the default option
+2. If the user-provided value is an object, a new `{}` object is created with values recursively merged
+3. Arrays and non-`undefined` literals -including `null`- are used as-is
+
+See [[Reference] feat: add meta.defaultOptions](https://github.com/eslint/eslint/pull/17656) > [`function deepMerge`](https://github.com/eslint/eslint/pull/17656/files#diff-9bbcc1ca9625b99554a055ac028190e5bd28432207a7f1a519690c5d0484287fR24) and [`describe("deepMerge")`](https://github.com/eslint/eslint/pull/17656/files#diff-fa70d7d4c142a6c3e727cf7280f981cb80a1db10bd6cc85c205e47e47da05cfeR25) for the proposed defaulting logic.
+
+### Rules with Differing Options Behavior
+
+[[Reference] feat: add meta.defaultOptions](https://github.com/eslint/eslint/pull/17656) implements `meta.defaultOptions` on rules that include options, are not deprecated, and match any of:
+
+- Are mentioned earlier in this RFC
+- Have a name starting with `A-C` and include the search string `context.options`
+- Include the search string `Object.assign(`
+- Experience test failures when Ajv's `useDefaults` is option is disabled (see [Schema Defaults](#schema-defaults))
+
+<details>
+<summary>
+The following matched rules will not use <code>meta.defaultOptions</code> because their current defaulting logic is incompatible with the new merging logic:
+</summary>
+
+- `array-bracket-newline`:
+  - Presumed `defaultOptions: [{ minItems: null, multiline: true }]`
+  - When given `options: [{ minItems: null }]`:
+    - Current normalized options: `[ { minItems: null } ]`, with implicit `multiline: false`
+    - Updated normalized options: `[ { minItems: null, multiline: true } ]`
+- `computed-property-spacing`:
+  - Presumed `defaultOptions: ["never"]`
+    - Note that `default: true` should also be removed from its `meta.schema`'s object `enforceForClassMembers` property
+  - When given `["always", {}]`:
+    - Current normalized `options.enforceForClassMembers`: `true`
+    - Updated normalized `options.enforceForClassMembers`: `undefined`
```

Comment:
Can `defaultOptions` for this rule be `["never", { enforceForClassMembers: true }]`?

Reply:
Oh! That didn't occur to me 😄 thanks!